File: cardmagic.cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long ll;

int M = 1000000009,N,K,T;

ll dp[101][5001];

ll solve(int index, int curr) {
    if (curr < 0) return 0;
    if (dp[index][curr] != -1) return dp[index][curr];
    if (index < N) {
        ll res = 0;
        for (int i = 1; i <= K; i++) res = (res + solve(index + 1, curr - i)) % M;
        dp[index][curr] = res;
    }
    return dp[index][curr];
}

int main() {
    scanf("%d %d %d",&N,&K,&T);
    memset(dp, -1, sizeof(dp));
    for (int i = 1; i < T + 1; i++) dp[N][i] = 0;
    dp[N][0] = 1;
    printf("%lld\n",solve(0,T));
    return 0;
}

File: cardmagic.py
from sys import setrecursionlimit

# use top down dynamic programming with memoisation in O(NT) time
setrecursionlimit(1000000)

M = pow(10,9) + 9
N,K,T = map(int,input().split())
dp = [[-1 for i in range(T + 1)] for j in range(N + 1)]
for i in range(1,T + 1):
    dp[N][i] = 0
dp[N][0] = 1

def solve(index,curr):
    if curr < 0:
        return 0
    if dp[index][curr] != -1:
        return dp[index][curr]
    if index < N:
        res = 0
        for i in range(1,K + 1):
            res += solve(index + 1, curr - i)
            res %= M
        dp[index][curr] = res
    return dp[index][curr]

print(solve(0,T))
